From 84a3c6759f15b7cedc4ed1fd292a64d91c0bd127 Mon Sep 17 00:00:00 2001
From: Rajat Soni <quic_rajson@quicinc.com>
Date: Thu, 8 Sep 2022 14:48:28 +0530
Subject: [PATCH] ath12k: delete desc list before adding to free list

Host crash is seen when we do rmmod after pinging AP to STA
because in ath12k_dp_cc_cleanup() during rx desc cleanup
dma_unmap_single is getting incorrect address that causes crash.

To resolve this desc info list should be deleted before adding
to free list.

Signed-off-by: Rajat Soni <quic_rajson@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index eada13a..978bc40 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3004,7 +3004,7 @@ try_again:
 		desc_info->skb = NULL;
 
 		spin_lock_bh(&dp->rx_desc_lock);
-		list_add_tail(&desc_info->list, &dp->rx_desc_free_list);
+		list_move_tail(&desc_info->list, &dp->rx_desc_free_list);
 		spin_unlock_bh(&dp->rx_desc_lock);
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
@@ -3702,7 +3702,7 @@ ath12k_dp_process_rx_err_buf(struct ath12k *ar, u32 *ring_desc,
 	msdu = desc_info->skb;
 	desc_info->skb = NULL;
 	spin_lock_bh(&ab->dp.rx_desc_lock);
-	list_add_tail(&desc_info->list, &ab->dp.rx_desc_free_list);
+	list_move_tail(&desc_info->list, &ab->dp.rx_desc_free_list);
 	spin_unlock_bh(&ab->dp.rx_desc_lock);
 
 	rxcb = ATH12K_SKB_RXCB(msdu);
@@ -4204,7 +4204,7 @@ int ath12k_dp_rx_process_wbm_err(struct ath12k_base *ab,
 		desc_info->skb = NULL;
 
 		spin_lock_bh(&dp->rx_desc_lock);
-		list_add_tail(&desc_info->list, &dp->rx_desc_free_list);
+		list_move_tail(&desc_info->list, &dp->rx_desc_free_list);
 		spin_unlock_bh(&dp->rx_desc_lock);
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
-- 
2.17.1

